### DISCUSSION TITLE
mTLS binding and direct grant

### DISCUSSION THREAD

**Discussion by bw512** [keycloak/keycloak#13482] [Q&A] [answered]
I have been trying out the "OAuth 2.0 Mutual TLS Certificate Bound Access Tokens Enabled" feature. When this is enabled, a browser based login flow with mTLS correctly adds the `cnf` field into the access token with the certificate hash. However, if I use a direct grant (example of what I'm doing below) then the `cnf` field does not seem to be added into the access token. Is this expected behavior, or have I perhaps missed some extra configuration steps?

---

I have setup keycloak to use a login flow which does an X509/Validate Username. I have "OAuth 2.0 Mutual TLS Certificate Bound Access Tokens Enabled" selected in the client configuration.

```bash
curl --cert <path-to-user-cert> --key <path-to-user-key> -X POST 'https://<keycloak-url>/protocol/openid-connect/token' \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data-urlencode 'client_id=<client-id>' \
  --data-urlencode 'grant_type=password'
```

**Comment by tnorimat** [comment, marked as answer]
Hello @bw512, 

The current keycloak (kc19) support this holder-of-key bound token for the following grant:
- authorization code
- client credentials
- device
- CIBA

The current keycloak (kc19) does not support the holder-of-key bound token for resource owner password credentials grant (direct grant in keycloak).

**Reply by bw512** [reply]
Thanks @tnorimat . Do you think this feature is likely to be added in the future?

**Comment by tnorimat** [comment]
@bw512 It is not 100% sure that this issue be treated but could you create a corresponding issue on the issue tab?

**Reply by bw512** [reply]
I have raised https://github.com/keycloak/keycloak/issues/13662

**Reply by tnorimat** [reply]
@bw512 the issue was resolved.

**Comment by tnorimat** [comment]
@bw512 It seems that you intended to bind a user (a user's private key) with an access token. Could you tell me which kind of use cases this user-bound token is used in?

**Reply by bw512** [reply]
Thanks. I'm happy to raise an issue, but let me explain the use-case first just in-case there are more appropriate solutions that negate the need.

**Background**
My understanding is that, in a web application context, mTLS binding can be used as follows:
* User visits web app
* Web app logs user in via authorisation code grant with PKCE (for example)
    * User is logged in to keycloak using mTLS credentials, users certificate hash is bound into access token
* Web app calls HTTP API using mTLS and provides access token from above
* HTTP API can validate access token & grants, as well as confirm users certificate hash (from mTLS) matches what's in the access token as proof of possession

**Use Case**
We'd like to use the proof of possession approach, but for a desktop application that is calling the HTTP API. The desktop app is trusted with the users TLS certificates to do mTLS. This could proceed as follows:
* User opens desktop app
* Desktop app uses password credentials grant to obtain access token bound to the TLS certificate hash
* Desktop app calls HTTP API using mTLS and provides access token from above
* HTTP API can validate access token & grants, as well as confirm users certificate hash (from mTLS) matches what's in the access token as proof of possession

**Reply by tnorimat** [reply]
@bw512 Thank you for the clarification.
